File: linked_list.js
class Node {
    constructor (value) {
        this.value = value
        this.next = undefined
    }   
}

class LinkedList {
    constructor () {
        var dummy = new Node("dummy")
        this.before = undefined
        this.current = undefined

        this.head = dummy
        this.tail = dummy
        
        this.num_of_data = 0
    }

    append (value) {
        var new_node = new Node(value)
        this.tail.next = new_node
        this.tail = new_node

        this.num_of_data += 1
    }

    delete () {
        let delete_node = this.current.value
        if(this.current == this.tail) {
            this.tail = this.before
        }

        this.before.next = this.current.next
        this.current = this.before

        this.num_of_data -= 1
        return delete_node
    }

    status() {
        console.log('---------STATUS---------')
        console.log(`current : ${this.current.value}`)
        console.log(`head : ${this.head.value}`)
        console.log(`tail : ${this.tail.value}`)
        console.log(`num_of_data : ${this.num_of_data}`)
    }

    print() {
        this.current = this.head.next
        let data = this.current
        if(data) {
            console.log(data.value)
            while (true) {
                data = this.next()
                if(data) {
                    console.log(data.value)
                } else {
                    break
                }
            }
        }
    }

    first() {
        if (this.num_of_data === 0 ){
            return undefined
        }
        this.before = this.head
        this.current = this.head.next

        return this.current.value
    }

    next() {
        if (this.current.next == undefined) {
            return undefined
        }
        this.before = this.current
        this.current = this.current.next
        return this.current
    }
}

var linked_list = new LinkedList()
linked_list.append(1)
linked_list.append(2)
linked_list.append(3)
linked_list.append(4)
linked_list.append(5)
linked_list.append(6)
linked_list.append(7)
linked_list.first()
linked_list.next()
linked_list.next()
linked_list.next()
linked_list.next()
linked_list.next()

linked_list.status()
linked_list.delete()
linked_list.status()

linked_list.print()

File: linked_list.py
class Node:
    def __init__(self, value) :
        self.value = value
        self.next = next
    
class LinkedList:
    def __init__(self) :
        dummy = Node("dummy")
        self.head = dummy
        self.tail = dummy

        self.current = None
        self.before = None

        self.num_of_data = 0

    def append(self, value) :
        new_node = Node(value)
        self.tail.next = new_node
        self.tail = new_node
        self.num_of_data += 1
    
    def delete(self) :
        pop_data = self.current.data
        if self.current is self.tail :
            self.tail = self.before 
        
        self.before.next = self.current.next
        self.current = self.before

        self.num_of_data -= 1
        return pop_data
    
    def first(self) :
        if self.num_of_data == 0:
            return None
        self.before = self.head
        self.current = self.head.next
        return self.current.data

    def next(self) :
        if self.current.next == None:
            return None
        self.before = self.current
        self.current = self.current.next

        return self.current.value
    
    def size(self) :
        return self.num_of_data

    def show(self) :
        temp = self.current
        

l_list = LinkedList()
l_list.append